---
title: "Metagenomics_analysis"
output: html_notebook
---

## Versions

R version: 4.2.0

jsonlite version: 1.8.4

dplyr version: 1.0.9

ggplot2 version: 3.3.6

## Data Preparations

The following section will prepare input data to be used in the analysis

```{r read data,warning=FALSE, message=FALSE}
library(jsonlite)
library(dplyr)
library(ggplot2)

# Sets current path as working file path for all code chuncks
knitr::opts_knit$set(root.dir = ".")

# Read JSON file with all the pathways and the enzymes within into R object
PW.all.enz <- fromJSON("data/all_pathways_enzymes_dict.json")
# Read JSON file with all the pathways codes and their names within into R object
PW.all.names <- fromJSON("data/ec_pathway_names_dict.json")

# Read  the metagenomics data file
mgxCount <- read.table(file = '../analysis/data/mgxData', sep = '\t', header = TRUE)
# Read metadata file sample labels
metaData <- read.table(file = "../analysis/data/metaData", sep = '\t', stringsAsFactors = TRUE, header = TRUE)

# Turns the first row of mgxData (Gene.Family) into row name
rownames(mgxCount) <- mgxCount[,1]
mgxCount <- mgxCount[, -1]

#metaData$diagnosis <- relevel(metaData$diagnosis,ref="nonIBD")
```

## Filtering Steps

We will apply some filtering process to filter out genes in the input data

```{r filtering,warning=FALSE, message=FALSE}
# Remove genes which has all zero values for all samples then start DE analysis
mgxCount[is.na(mgxCount)] <- 0
nonzero <- rowSums(mgxCount) > 0
mgxCount <- mgxCount[nonzero,]

remove(nonzero)
```

## T-tests

```{r}
# Separate CD, UC and nonIBD
metaData.CD <- metaData[metaData$diagnosis=="CD",]
metaData.UC <- metaData[metaData$diagnosis=="UC",]
metaData.nonIBD <- metaData[metaData$diagnosis=="nonIBD",]

# Select metagenomics data from CD, UC and nonIBD
mgxCount.CD <- subset(mgxCount, select=metaData.CD$External.ID)
mgxCount.UC <- subset(mgxCount, select=metaData.UC$External.ID)
mgxCount.nonIBD <- subset(mgxCount, select=metaData.nonIBD$External.ID)

#remove(metaData.CD, metaData.UC, metaData.nonIBD, metaData, mgxCount)
```

### T-tests for Crohn's Disease (CD)

```{r}
# Compute statistical significance (using t-test)
pValue.CD = NULL # Empty list for the p-values

for(i in 1 : nrow(mgxCount.nonIBD)) { # For each gene : 
	x = mgxCount.nonIBD[i,] # control of gene number i
	y = mgxCount.CD[i,] # CD of gene number i
	
	# Compute t-test between the two conditions
	t = t.test(x, y)
	
	# Put the current p-value in the pvalues list
	pValue.CD[i] = t$p.value
}

# Apply Benjamini Hochberg correction to pvalues to correct for multiple testing
pAdjust.CD = p.adjust(pValue.CD, method="BH", n= nrow(mgxCount.nonIBD))

#remove(pValue.CD, t, x, y, i)
```

```{r}
library("DESeq2")

metaData.nonIBD.CD <- metaData[(metaData$diagnosis=="CD" | metaData$diagnosis=="nonIBD"),]
mgxCount.nonIBD.CD <- subset(mgxCount, select=metaData.nonIBD.CD$External.ID)

dds.CD <- DESeqDataSetFromMatrix(countData=round(mgxCount.nonIBD.CD), 
                              colData=metaData.nonIBD.CD, 
                              design=~diagnosis, tidy = FALSE)

dds.CD <- DESeq(dds.CD)

res.CD <- results(dds.CD)
summary(res.CD)

n_distinct(which(res.CD$pvalue<0.05))
n_distinct(which(res.CD$padj<0.05))
```


### T-tests for Ulcerative colitis (UC)

```{r}
# Compute statistical significance (using t-test)
pValue.UC = NULL # Empty list for the p-values

for(i in 1 : nrow(mgxCount.nonIBD)) { # For each gene : 
	x = mgxCount.nonIBD[i,] # control of gene number i
	y = mgxCount.UC[i,] # UC of gene number i
	
	# Compute t-test between the two conditions
	t = t.test(x, y)
	
	# Put the current p-value in the pvalues list
	pValue.UC[i] = t$p.value
}

# Apply Benjamini Hochberg correction to pvalues to correct for multiple testing
pAdjust.UC = p.adjust(pValue.UC, method="BH", n= nrow(mgxCount.nonIBD))

remove(pValue.UC, t, x, y, i  )#,mgxCount.nonIBD)
```

```{r}
metaData.nonIBD.UC <- metaData[(metaData$diagnosis=="UC" | metaData$diagnosis=="nonIBD"),]
mgxCount.nonIBD.UC <- subset(mgxCount, select=metaData.nonIBD.UC$External.ID)

dds.UC <- DESeqDataSetFromMatrix(countData=round(mgxCount.nonIBD.UC), 
                              colData=metaData.nonIBD.UC, 
                              design=~diagnosis, tidy = FALSE)

dds.UC <- DESeq(dds.UC)

res.UC <- results(dds.UC)
summary(res.UC)

n_distinct(which(res.UC$pvalue<0.05))
n_distinct(which(res.UC$padj<0.05))
```


### Select statistically significant enzymes

```{r}
# Select statistically significant enzymes
deg.CD <- mgxCount.CD[which(pAdjust.CD<0.05),]
deg.CD$pAdjust <- pAdjust.CD[which(pAdjust.CD<0.05)]
deg.UC <- mgxCount.UC[which(pAdjust.UC<0.05),]
deg.UC$pAdjust <- pAdjust.UC[which(pAdjust.UC<0.05)]

# Get the row names and split them at the colon
deg.UC.EC <- row.names(deg.UC)
deg.UC.EC <- lapply(strsplit(deg.UC.EC, ":"), "[[", 1)
deg.CD.EC <- row.names(deg.CD)
deg.CD.EC <- lapply(strsplit(deg.CD.EC, ":"), "[[", 1)

remove(pAdjust.CD, pAdjust.UC)
```

## Calculate the ORA score for each pathway, using the Fishers exact test

Based on: <https://www.pathwaycommons.org/guide/primers/statistics/fishers_exact_test/>

### ORA score for Ulcerative colitis (UC)


```{r}
library("clusterProfiler")

deg.nonIBD.EC <- row.names(mgxCount.nonIBD)
deg.nonIBD.EC <- lapply(strsplit(deg.nonIBD.EC, ":"), "[[", 1)

PW.to.enzyme <- as.data.frame(t(as.data.frame(sapply(PW.all.enz, rbind))))
PW.to.enzyme$pathway <- rownames(PW.to.enzyme)
colnames(PW.to.enzyme) <- c("enzyme", "pathway")
PW.to.enzyme <- PW.to.enzyme[,c(2,1)]
PW.to.enzyme$pathway <- sub("\\..*", "", PW.to.enzyme$pathway)

PW.to.name <- data.frame("pathway"=names(PW.all.names), "name"=unlist(PW.all.names))

res <- enricher(gene=as.factor(unlist(deg.UC.EC)),
                TERM2GENE = PW.to.enzyme,
                TERM2NAME = PW.to.name)

summary(res)
```


```{r}
#Create a dataframe to store the required numbers in.
Contingency_table.UC <- data.frame(matrix(ncol=5,nrow=0, dimnames=list(NULL, c("EC.PW", "x", "m", "n", "k"))))
counter = 1
for (i in 1:length(PW.all.enz)) {
   Contingency_table.UC[counter,1] <- names(PW.all.enz[i]) #EC.PW
   Contingency_table.UC[counter,2] <- length(intersect(deg.UC.EC, PW.all.enz[[i]])) ##x <- (number4) #Total differentially changed enzymes, also in a PW.
   Contingency_table.UC[counter,3] <- length(PW.all.enz[[i]]) ##m <- (number) #Total Enzymes in PW
   Contingency_table.UC[counter,4] <- nrow(mgxCount.UC) - length(intersect(deg.UC.EC, PW.all.enz[[i]])) ##n <- (number2) #Total Metabolites measured not in PW
   Contingency_table.UC[counter,5] <- nrow(deg.UC) ##k <- (number3) #Total differentially changed enzymes

   counter <- counter + 1
}

# Calculate hypergeometric density p-value for all pathways.
i <- 1:nrow(Contingency_table.UC)
probabilities.UC <- dhyper(Contingency_table.UC[i,2], Contingency_table.UC[i,3], Contingency_table.UC[i,4], Contingency_table.UC[i,5], log = FALSE)

remove(mgxCount.UC, Contingency_table.UC, deg.UC, deg.UC.EC, counter, i)
```

### ORA score for Crohn's Disease (CD)

```{r}
res <- enricher(gene=as.factor(unlist(deg.CD.EC)),
                TERM2GENE = PW.to.enzyme,
                TERM2NAME = PW.to.name)

summary(res)
```


```{r}
#Create a dataframe to store the required numbers in.
Contingency_table.CD <- data.frame(matrix(ncol=5,nrow=0, dimnames=list(NULL, c("EC.PW", "x", "m", "n", "k"))))
counter = 1
for (i in 1:length(PW.all.enz)) {
   Contingency_table.CD[counter,1] <- names(PW.all.enz[i]) #EC.PW
   Contingency_table.CD[counter,2] <- length(intersect(deg.CD.EC, PW.all.enz[[i]])) ##x <- (number4) #Total differentially changed enzymes, also in a PW.
   Contingency_table.CD[counter,3] <- length(PW.all.enz[[i]]) ##m <- (number) #Total Enzymes in PW
   Contingency_table.CD[counter,4] <- nrow(mgxCount.CD) - length(intersect(deg.CD.EC, PW.all.enz[[i]])) ##n <- (number2) #Total Metabolites measured not in PW
   Contingency_table.CD[counter,5] <- nrow(deg.CD) ##k <- (number3) #Total differentially changed enzymes

   counter <- counter + 1
}

# Calculate hypergeometric density p-value for all pathways.
i <- 1:nrow(Contingency_table.CD)
probabilities.CD <- dhyper(Contingency_table.CD[i,2], Contingency_table.CD[i,3], Contingency_table.CD[i,4], Contingency_table.CD[i,5], log = FALSE)

remove(mgxCount.CD, Contingency_table.CD, deg.CD, deg.CD.EC, counter, i)
```

## Plot the results in a horizontal bar chart

### Plot for Ulcerative colitis (UC)

```{r}
PW.all.names <- as.matrix(PW.all.names)

# Create a dataframe for the results of the pathway analysis
PW_Analysis_results.UC <- data.frame(names(PW.all.enz), PW.all.names, probabilities.UC, lengths(PW.all.enz))
colnames(PW_Analysis_results.UC) <- c("EC_code", "name", "probabilities", "count")

# Order pathway analysis results by p-values
PWAnalysis_results_sorted.UC <- PW_Analysis_results.UC[with(PW_Analysis_results.UC, order(probabilities)),]

remove(PW_Analysis_results.UC, probabilities.UC)
```

```{r fig.height = 15, fig.width = 9}
# Horizontal bar plot of significant pathways in UC
PWAnalysis_results_sorted.UC[which(PWAnalysis_results_sorted.UC$probabilities < 0.05),] %>% 
  ggplot(aes(reorder(as.character(name), count), count)) + 
  geom_col(aes(fill = probabilities)) + 
  scale_fill_gradient2(low = "#0000ff", 
                       mid = "#ff0000",
                       high= "yellow",
                       midpoint = 0.01) +
  coord_flip() + 
  labs(x = "Pathways (UC)", y = "Enzyme count")
```

### Plot for Crohn's Disease (CD)

```{r}
# Create a dataframe for the results of the pathway analysis
PW_Analysis_results.CD <- data.frame(names(PW.all.enz), PW.all.names, probabilities.CD, lengths(PW.all.enz))
colnames(PW_Analysis_results.CD) <- c("EC_code", "name", "probabilities", "count")

# Order pathway analysis results by p-values
PWAnalysis_results_sorted.CD <- PW_Analysis_results.CD[with(PW_Analysis_results.CD, order(probabilities)),]

remove(PW_Analysis_results.CD, probabilities.CD, PW.all.names, PW.all.enz)
```

```{r fig.height = 15, fig.width = 9}
# Horizontal bar plot of significant pathways in CD
PWAnalysis_results_sorted.CD[which(PWAnalysis_results_sorted.CD$probabilities < 0.05),] %>% 
  ggplot(aes(reorder(as.character(name), count), count)) + 
  geom_col(aes(fill = probabilities)) + 
  scale_fill_gradient2(low = "#0000ff", 
                       mid = "#ff0000",
                       high= "yellow",
                       midpoint = 0.01) +
  coord_flip() + 
  labs(x = "Pathways (CD)", y = "Enzyme count")
```

## Writing output tables in csv file

```{r}
# Write the output in a file
PWAnalysis_results_sorted.UC <- apply(PWAnalysis_results_sorted.UC,2,as.character)
write.table(PWAnalysis_results_sorted.UC, "output/mgxPWdata_UC.csv", sep =",", row.names = FALSE)

# Write the output in a file
PWAnalysis_results_sorted.CD <- apply(PWAnalysis_results_sorted.CD,2,as.character)
write.table(PWAnalysis_results_sorted.CD, "output/mgxPWdata_CD.csv", sep =",", row.names = FALSE)
```
