---
title: "Metagenomics_analysis"
output: html_notebook
---

## Versions

R version: 4.2.0

jsonlite version: 1.8.4

dplyr version: 1.0.9

ggplot2 version: 3.3.6


## Data Preparations

The following section will prepare input data to be used in the analysis

```{r read data,warning=FALSE, message=FALSE}
library(jsonlite)
library(dplyr)
library(ggplot2)

# Sets current path as working file path for all code chuncks
knitr::opts_knit$set(root.dir = ".")

# Read JSON file with all the pathways and the enzymes within into R object
pathways.all <- fromJSON("data/all_pathways_enzymes_dict.json")
# Read JSON file with all the pathways codes and their names within into R object
pathways.all.names <- fromJSON("data/ec_pathway_names_dict.json")

# Read  the metagenomics data file
mgxCount <- read.table(file = '../analysis/data/mgxData', sep = '\t', header = TRUE)
# Read metadata file sample labels
metaData <- read.table(file = "../analysis/data/metaData", sep = '\t', stringsAsFactors = TRUE, header = TRUE)

# Turns the first row of mgxData (Gene.Family) into row name
rownames(mgxCount) <- mgxCount[,1]
mgxCount <- mgxCount[, -1]

#metaData$diagnosis <- relevel(metaData$diagnosis,ref="nonIBD")
```

## Filtering Steps

We will apply some filtering process to filter out genes in the input data

```{r filtering,warning=FALSE, message=FALSE}
# Remove genes which has all zero values for all samples then start DE analysis
mgxCount[is.na(mgxCount)] <- 0
nonzero <- rowSums(mgxCount) > 0
mgxCount <- mgxCount[nonzero,]

remove(nonzero)
```

## T-tests

```{r}
# Separate CD, UC and nonIBD
metaData.CD <- metaData[metaData$diagnosis=="CD",]
metaData.UC <- metaData[metaData$diagnosis=="UC",]
metaData.nonIBD <- metaData[metaData$diagnosis=="nonIBD",]

# Select metagenomics data from CD, UC and nonIBD
mgxCount.CD <- subset(mgxCount, select=metaData.CD$External.ID)
mgxCount.UC <- subset(mgxCount, select=metaData.UC$External.ID)
mgxCount.nonIBD <- subset(mgxCount, select=metaData.nonIBD$External.ID)

remove(metaData.CD, metaData.UC, metaData.nonIBD)
```

### T-tests for Crohn's Disease (CD)

```{r}
# Compute statistical significance (using t-test)
pValue.CD = NULL # Empty list for the p-values
tStat.CD = NULL # Empty list of the t test statistics

for(i in 1 : nrow(mgxCount.nonIBD)) { # For each gene : 
	x = mgxCount.nonIBD[i,] # control of gene number i
	y = mgxCount.CD[i,] # CD of gene number i
	
	# Compute t-test between the two conditions
	t = t.test(x, y)
	
	# Put the current p-value in the pvalues list
	pValue.CD[i] = t$p.value
}
pAdjust.CD = p.adjust(pValue.CD, method="BH", n= nrow(mgxCount.nonIBD))
```

### T-tests for Ulerative colitis (UC)

```{r}
# Compute statistical significance (using t-test)
pValue.UC = NULL # Empty list for the p-values
tStat.UC = NULL # Empty list of the t test statistics

for(i in 1 : nrow(mgxCount.nonIBD)) { # For each gene : 
	x = mgxCount.nonIBD[i,] # control of gene number i
	y = mgxCount.UC[i,] # UC of gene number i
	
	# Compute t-test between the two conditions
	t = t.test(x, y)
	
	# Put the current p-value in the pvalues list
	pValue.UC[i] = t$p.value
}

remove(t, x, y, i)
pAdjust.UC = p.adjust(pValue.UC, method="BH")
```

```{r}
# Select statistically significant enzymes
deg.CD <- mgxCount.CD[which(pValue.CD<0.05),]
deg.CD$pvalue <- pValue.CD[which(pValue.CD<0.05)]
deg.UC <- mgxCount.UC[which(pValue.UC<0.05),]
deg.UC$pvalue <- pValue.UC[which(pValue.UC<0.05)]

remove(pValue.CD, pValue.UC)
```

```{r}
# Select statistically significant enzymes
deg.CD <- mgxCount.CD[which(pAdjust.CD<0.05),]
deg.CD$pAdjust <- pAdjust.CD[which(pAdjust.CD<0.05)]
deg.UC <- mgxCount.UC[which(pAdjust.UC<0.05),]
deg.UC$pAdjust <- pAdjust.UC[which(pAdjust.UC<0.05)]

remove(pValue.CD, pValue.UC)
```

# Calculate the ORA score for each pathway, using the Fishers exact test.

## ORA score for Ulerative colitis (UC)

```{r}
##Based on: https://www.pathwaycommons.org/guide/primers/statistics/fishers_exact_test/

# Get the row names and split them at the colon
deg.UC.EC <- row.names(deg.UC)
deg.UC.EC <- lapply(strsplit(deg.UC.EC, ":"), "[[", 1)

#Create a dataframe to store the required numbers in.
Contingency_table.UC <- data.frame(matrix(ncol=5,nrow=0, dimnames=list(NULL, c("EC.PW", "x", "m", "n", "k"))))
counter = 1
for (i in 1:length(pathways.all)) {
   Contingency_table.UC[counter,1] <- names(pathways.all[i]) #EC.PW
   Contingency_table.UC[counter,2] <- length(intersect(deg.UC.EC, pathways.all[[i]])) ##x <- (number4) #Total differentially changed enzymes, also in a PW.
   Contingency_table.UC[counter,3] <- length(pathways.all[[i]]) ##m <- (number) #Total Enzymes in PW
   Contingency_table.UC[counter,4] <- nrow(mgxCount.UC) - length(intersect(deg.UC.EC, pathways.all[[i]])) ##n <- (number2) #Total Metabolites measured not in PW
   Contingency_table.UC[counter,5] <- nrow(deg.UC) ##k <- (number3) #Total differentially changed enzymes

   counter <- counter + 1
}

# Calculate hypergeometric density p-value for all pathways.
i <- 1:nrow(Contingency_table.UC)
probabilities.UC <- dhyper(Contingency_table.UC[i,2], Contingency_table.UC[i,3], Contingency_table.UC[i,4], Contingency_table.UC[i,5], log = FALSE)

```

## ORA score for Crohn's Disease (CD)

```{r}
# Get the row names and split them at the colon
deg.CD.EC <- row.names(mgxCount.CD)
deg.CD.UC <- lapply(strsplit(deg.CD.EC, ":"), "[[", 1)


Contingency_table.CD <- data.frame(matrix(ncol=5,nrow=0, dimnames=list(NULL, c("EC.PW", "x", "m", "n", "k"))))
counter = 1
for (i in 1:length(pathways.all)) {
   Contingency_table.CD[counter,1] <- names(pathways.all[i]) #EC.PW
   Contingency_table.CD[counter,2] <- length(intersect(deg.CD.EC, pathways.all[[i]])) ##x <- (number4) #Total differentially changed enzymes, also in a PW.
   Contingency_table.CD[counter,3] <- length(pathways.all[[i]]) ##m <- (number) #Total Enzymes in PW
   Contingency_table.CD[counter,4] <- nrow(mgxCount.CD) - length(intersect(deg.CD.EC, pathways.all[[i]])) ##n <- (number2) #Total Metabolites measured not in PW
   Contingency_table.CD[counter,5] <- nrow(deg.CD) ##k <- (number3) #Total differentially changed enzymes

   counter <- counter + 1
}

# Calculate hypergeometric density p-value for all pathways.
i <- 1:nrow(Contingency_table.CD)
probabilities.CD <- dhyper(Contingency_table.CD[i,2], Contingency_table.CD[i,3], Contingency_table.CD[i,4], Contingency_table.CD[i,5], log = FALSE)
```

# Plot the results in a horizontal bar chart

## Plot for Ulerative colitis (UC)

```{r}
#pathways_names$code <- rownames(pathways.all.names)
#all(rownames(pathwayAnalysis_results) == rownames(pathways.all.names))
pathways_names <- as.matrix(pathways.all.names)

pathwayAnalysis_results.UC <- data.frame(names(pathways.all), pathways_names, probabilities.UC, lengths(pathways.all))
colnames(pathwayAnalysis_results.UC) <- c("EC_code", "name", "probabilities", "count")

pathwayAnalysis_results_sorted.UC <- pathwayAnalysis_results.UC[with(pathwayAnalysis_results.UC, order(probabilities)),]

# Write the output in a file
pathwayAnalysis_results_sorted.UC <- apply(pathwayAnalysis_results_sorted.UC,2,as.character)
write.table(pathwayAnalysis_results_sorted.UC, "output/mgxPWdata_UC.csv", sep =",", row.names = FALSE)
```

```{r fig.height = 15, fig.width = 9}
pathwayAnalysis_results_sorted.UC[which(pathwayAnalysis_results_sorted.UC$probabilities < 0.05),] %>% 
  ggplot(aes(reorder(as.character(name), count), count)) + 
  geom_col(aes(fill = probabilities)) + 
  scale_fill_gradient2(low = "#0000ff", 
                       mid = "#ff0000",
                       high= "yellow",
                       midpoint = 0.01) +
  coord_flip() + 
  labs(x = "Pathways (UC)", y = "Enzyme count")
```

## Plot for Crohn's Disease (CD)

```{r}
#pathways_names$code <- rownames(pathways.all.names)
#all(rownames(pathwayAnalysis_results) == rownames(pathways.all.names))
#pathways_names <- as.matrix(pathways.all.names)

pathwayAnalysis_results.CD <- data.frame(names(pathways.all), pathways_names, probabilities.CD, lengths(pathways.all))
colnames(pathwayAnalysis_results.CD) <- c("EC_code", "name", "probabilities", "count")

pathwayAnalysis_results_sorted.CD <- pathwayAnalysis_results.UC[with(pathwayAnalysis_results.UC, order(probabilities)),]

# Write the output in a file
pathwayAnalysis_results_sorted.CD <- apply(pathwayAnalysis_results_sorted.CD,2,as.character)
write.table(pathwayAnalysis_results_sorted.CD, "output/mgxPWdata_CD.csv", sep =",", row.names = FALSE)
```

```{r fig.height = 15, fig.width = 9}
pathwayAnalysis_results_sorted.CD[which(pathwayAnalysis_results_sorted.CD$probabilities < 0.05),] %>% 
  ggplot(aes(reorder(as.character(name), count), count)) + 
  geom_col(aes(fill = probabilities)) + 
  scale_fill_gradient2(low = "#0000ff", 
                       mid = "#ff0000",
                       high= "yellow",
                       midpoint = 0.01) +
  coord_flip() + 
  labs(x = "Pathways (CD)", y = "Enzyme count")
```
