---
title: "Metabolomics_analysis"
output: html_document
---

## Libraries

```{r, message=FALSE, warning=FALSE}
# Check if libraries are already installed > otherwise install it
if(!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager",repos = "http://cran.us.r-project.org")
if(!"knitr" %in% installed.packages()) BiocManager::install("knitr")
if(!"jsonlite" %in% installed.packages()) BiocManager::install("jsonlite")
if(!"dplyr" %in% installed.packages()) BiocManager::install("dplyr")
if(!"clusterProfiler" %in% installed.packages()) BiocManager::install("clusterProfiler")
if(!"ggplot2" %in% installed.packages()) BiocManager::install("ggplot2")

# Load packages
library(knitr)
library(jsonlite)
library(dplyr)
library(clusterProfiler)
library(ggplot2)
```

## Generate HMDB.csv for data annotation

This can be skipped...

```{r setup}
# Sets current path as working file path for all code chuncks
knitr::opts_knit$set(root.dir = ".")

# Read  the metabolomics data file
mbxCount <- read.table(file = "../preprocessing/output/mbxData", sep = '\t', header = TRUE)
# Read metadata file sample labels
metaData <- read.table(file = "../preprocessing/output/mbxMetaData", sep = '\t', stringsAsFactors = TRUE, header = TRUE)

# Read JSON file with all the pathways and the compounds within into R object
PW.to.compound <- fromJSON("data/all_pathways_compounds_dict.json")
# Read JSON file with all the pathways codes and their names within into R object
PW.all.names <- fromJSON("data/ec_pathway_names_dict.json")

```
## Filtering Steps

We will apply some filtering process to filter out genes in the input data

```{r filtering,warning=FALSE, message=FALSE}
# Remove genes which has all zero values for all samples then start DE analysis
mbxCount[is.na(mbxCount)] <- 0
nonzero <- rowSums(mbxCount[,c(3:ncol(mbxCount))]) > 0
mbxCount <- mbxCount[nonzero,]

remove(nonzero)
```

## T-tests

```{r}
# Separate CD, UC and nonIBD
metaData.CD <- metaData[metaData$diagnosis=="CD",]
metaData.UC <- metaData[metaData$diagnosis=="UC",]
metaData.nonIBD <- metaData[metaData$diagnosis=="nonIBD",]

# Select metagenomics data from CD, UC and nonIBD
mbxCount.CD <- subset(mbxCount, select=metaData.CD$External.ID)
mbxCount.UC <- subset(mbxCount, select=c(1,2, metaData.UC$External.ID))
mbxCount.nonIBD <- subset(mbxCount, select=c(1,2, metaData.nonIBD$External.ID))

#remove(metaData.CD, metaData.UC, metaData.nonIBD, metaData, mbxCount)
```

### T-tests for Crohn's Disease (CD)

```{r}
# Compute statistical significance (using t-test)
pValue.CD = NULL # Empty list for the p-values

for(i in 1 : nrow(mbxCount.nonIBD)) { # For each metabolite : 
	x = mbxCount.nonIBD[i,-(1:2)] # control of metabolite number i
	y = mbxCount.CD[i,-(1:2)] # CD of metabolite number i
	
	# Compute t-test between the two conditions
	t = t.test(x, y)
	
	# Put the current p-value in the pvalues list
	pValue.CD[i] = t$p.value
}

# Apply Benjamini Hochberg correction to pvalues to correct for multiple testing
pAdjust.CD = p.adjust(pValue.CD, method="BH", n= nrow(mbxCount.nonIBD))

#remove(pValue.CD, t, x, y, i)
```

```{r eval=FALSE}
library("DESeq2")

metaData.nonIBD.CD <- metaData[(metaData$diagnosis=="CD" | metaData$diagnosis=="nonIBD"),]
mbxCount.nonIBD.CD <- subset(mbxCount, select=metaData.nonIBD.CD$External.ID)

dds.CD <- DESeqDataSetFromMatrix(countData=mbxCount.nonIBD.CD[, c(3:ncol(mbxCount.nonIBD.CD))], 
                              colData=metaData.nonIBD.CD[,c(1, 3)], 
                              design=~diagnosis, tidy = FALSE)

dds.CD <- DESeq(dds.CD)

res.CD <- results(dds.CD)
summary(res.CD)

n_distinct(which(res.CD$pvalue<0.05))
n_distinct(which(res.CD$padj<0.05))
```


### T-tests for Ulcerative colitis (UC)

```{r}
# Compute statistical significance (using t-test)
pValue.UC = NULL # Empty list for the p-values

for(i in 1 : nrow(mbxCount.nonIBD)) { # For each gene : 
	x = mbxCount.nonIBD[i,-(1:2)] # control of gene number i
	y = mbxCount.UC[i,-(1:2)] # UC of gene number i
	
	# Compute t-test between the two conditions
	t = t.test(x, y)
	
	# Put the current p-value in the pvalues list
	pValue.UC[i] = t$p.value
}

# Apply Benjamini Hochberg correction to pvalues to correct for multiple testing
pAdjust.UC = p.adjust(pValue.UC, method="BH", n= nrow(mbxCount.nonIBD))

#remove(pValue.UC, t, x, y, i)
```


### Select statistically significant enzymes

```{r}
# Select statistically significant enzymes
deg.CD <- mbxCount.CD[which(pValue.CD<0.05),]
deg.CD$pValue <- pValue.CD[which(pValue.CD<0.05)]
deg.UC <- mbxCount.UC[which(pValue.UC<0.05),]
deg.UC$pValue <- pValue.UC[which(pValue.UC<0.05)]

#remove(pAdjust.CD, pAdjust.UC)
```

## Generate mbxData

This can be skipped...

https://tabbassidaloii.shinyapps.io/BridgeDb-Shiny/


## ...

```{r}
# Read metadata file sample labels
#metaData <- read.table(file = "../preprocessing/output/mbxMetaData", sep = '\t', stringsAsFactors = TRUE, header = TRUE)
# Read KEGG compound to HMDB list
compound.HMDB <- read.table(file = "data/XrefBatch_mapping_BridgeDB-Shiny.csv", sep = ',', header = TRUE)
# Remove ... columns
compound.HMDB <- compound.HMDB[, -c(2,4)]
# Change column names
colnames(compound.HMDB) <- c("HMDB", "compound")
```

```{r}
# Merge the KEGG compound to the full metabolomics data
deg.CD <- merge(deg.CD, compound.HMDB, by = "HMDB", all.x = TRUE)
length(which(is.na(deg.CD)))
# remove NAs for compounds
deg.CD <- deg.CD[!is.na(deg.CD$compound), ]
# Relocate the compound column as first
deg.CD <- deg.CD %>% relocate(compound)

# Merge the KEGG compound to the full metabolomics data
deg.UC <- merge(deg.UC, compound.HMDB, by = "HMDB", all.x = TRUE)
length(which(is.na(deg.UC)))
# remove NAs for compounds
deg.UC <- deg.UC[!is.na(deg.UC$compound), ]
# Relocate the compound column as first
deg.UC <- deg.UC %>% relocate(compound)

```

## Calculate the ORA score for each pathway


### ORA score for Ulcerative colitis (UC)


```{r}

PW.to.compound <- as.data.frame(t(as.data.frame(sapply(PW.to.compound, rbind))))
PW.to.compound$pathway <- rownames(PW.to.compound)
colnames(PW.to.compound) <- c("compound", "pathway")
PW.to.compound <- PW.to.compound[,c(2,1)]
PW.to.compound$pathway <- sub("\\..*", "", PW.to.compound$pathway)
PW.to.compound$compound <- sub("\\s\\s.*", "", PW.to.compound$compound)

PW.to.name <- data.frame("pathway"=names(PW.all.names), "name"=unlist(PW.all.names))

res.UC <- enricher(gene=as.factor(unlist(deg.UC)),
                TERM2GENE = PW.to.compound,
                TERM2NAME = PW.to.name)

summary(res.UC)
```


### ORA score for Crohn's Disease (CD)

```{r}
res.CD <- enricher(gene=as.factor(unlist(deg.CD)),
                TERM2GENE = PW.to.compound,
                TERM2NAME = PW.to.name)

summary(res.CD)
```

## Plot the results in a horizontal bar chart

### Plot for Ulcerative colitis (UC)


```{r}
# Horizontal bar plot of significant pathways in UC
res.UC[which(res.UC$p.adjust < 0.05),] %>% 
  ggplot(aes(reorder(as.character(Description), Count), Count)) + 
  geom_col(aes(fill = p.adjust)) + 
  scale_fill_gradient2(low = "#0000ff", 
                       mid = "#ff0000",
                       high= "yellow",
                       midpoint = 0.001) +
  coord_flip() + 
  labs(x = "Pathways (UC)", y = "Enzyme count")
```


### Plot for Crohn's Disease (CD)

```{r}
res.CD[which(res.CD$p.adjust < 0.05),] %>% 
  ggplot(aes(reorder(as.character(Description), Count), Count)) + 
  geom_col(aes(fill = p.adjust)) + 
  scale_fill_gradient2(low = "#0000ff", 
                       mid = "#ff0000",
                       high= "yellow",
                       midpoint = 0.001) +
  coord_flip() + 
  labs(x = "Pathways (UC)", y = "Enzyme count")
```

## Writing output tables in csv file

```{r}
# Write the output in a file
write.table(res.UC, "output/mbxPWdata_UC.csv", sep =",", row.names = FALSE)

# Write the output in a file
write.table(res.CD, "output/mbxPWdata_CD.csv", sep =",", row.names = FALSE)
```


## Print session info:

```{r print_session_info}
##Print session info:
sessionInfo()
```
